### PR TITLE
Adding VHost based org support

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Purpose
Fixes: https://github.com/wso2/apk/issues/663

### CODE REVIEW

**Comment by CrowleyRajapakse** [comment]
Updated Config with 3 dynamic listeners and related 3 dynamic routes: [config_dump1.txt](https://github.com/wso2/apk/files/11000648/config_dump1.txt)

File: helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml [hunk 1]
```diff
@@ -168,7 +170,7 @@ spec:
             - name: ROUTER_CLUSTER
               value: "default_cluster"
             - name: ROUTER_LABEL
-              value: "Default"
+              value: "default"
```
Comment (AmaliMatharaarachchi) [inline helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml:173]: don't we need to parameterize all label envs to `.Values.wso2.apk.dp.gateway.name`

File: adapter/config/parser.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ var (
 )
 
 // DefaultGatewayName represents the name of the default gateway
-const DefaultGatewayName = "Default"
+const DefaultGatewayName = "default"
```
Comment (AmaliMatharaarachchi) [inline adapter/config/parser.go:45]: does this have any usages now? shall we remove this from the code?

File: adapter/config/default_config.go [hunk 3]
```diff
@@ -35,7 +35,7 @@ var defaultConfig = &Config{
 		},
 		VhostMapping: []vhostMapping{
 			{
-				Environment: "Default",
+				Environment: "default",
```
Comment (AmaliMatharaarachchi) [inline adapter/config/default_config.go:38]: shall we remove VhostMapping as we get it now from gateway cr.

File: test/integration/integration/utils/kubernetes/helpers.go [hunk 4]
```diff
@@ -95,6 +95,40 @@ func WaitForGatewayAddress(t *testing.T, c client.Client, timeoutConfig config.T
 	return net.JoinHostPort(ipAddr, port)
 }
 
+// WaitForAPIListenerAddress waits until at least one IP Address has been set in the
+// Gateway infra exposed service.
+func WaitForAPIListenerAddress(t *testing.T, c client.Client, timeoutConfig config.TimeoutConfig) string {
```
Comment (Amila-Rukshan) [inline test/integration/integration/utils/kubernetes/helpers.go:100]: Shall we move the common logic in both `WaitForGatewayAddress` and `WaitForAPIListenerAddress` to another function and call it with lister port from two of these functions?

File: test/integration/integration/utils/http/http.go [hunk 5]
```diff
@@ -383,11 +383,21 @@ func GetTestToken(t *testing.T, gwAddr string, scopes ...string) string {
 	}
 	resp, err := client.Do(req)
 	if err != nil {
-		t.Fatalf("failed to get token: %v", err)
+		t.Logf("failed to get token: %v retrying after 20s ...", err)
```
Comment (Amila-Rukshan) [inline test/integration/integration/utils/http/http.go:386]: Is this to wait for enforcer test token endpoint to become available?

File: helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml [hunk 6]
```diff
@@ -149,6 +149,8 @@ spec:
               protocol: "TCP"
             - containerPort: 9090
               protocol: "TCP"
+            - containerPort: 9091
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml:152]: yes, fixed

File: helm-charts/templates/data-plane/gateway-components/gateway-runtime/router-service.yaml [hunk 7]
```diff
@@ -35,4 +35,7 @@ spec:
     - name: endpoint3
       protocol: TCP
       port: 9000
+    - name: endpoint4
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/gateway-runtime/router-service.yaml:38]: fixed

File: helm-charts/templates/data-plane/gateway-components/log-conf.yaml [hunk 8]
```diff
@@ -23,7 +23,7 @@ data:
     # 
     # default vhosts mapping for standalone mode
     [[adapter.vhostMapping]]
-      environment = "Default"
+      environment = "default"
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/log-conf.yaml]: no, removed and fixed

File: helm-charts/templates/data-plane/gateway-components/log-conf.yaml [hunk 9]
```diff
@@ -171,7 +171,7 @@ data:
     # Default virtual host mapping for standalone mode
     [[adapter.vhostMapping]]
     # The Adapter environment
-    environment = "Default"
+    environment = "default"
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/log-conf.yaml]: removed and fixed

File: helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml (hunk 1 above)
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml]: resolved

File: developer/tryout/samples/sample-api.yaml [hunk 10]
```diff
@@ -119,7 +122,9 @@ spec:
   parentRefs:
     - group: gateway.networking.k8s.io
       kind: Gateway
-      name: Default
+      name: default
+      sectionName: httpslistener
+      port: 9095
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply developer/tryout/samples/sample-api.yaml]: no, removed the port and resolved

File: adapter/config/parser.go (hunk 2 above)
Comment (CrowleyRajapakse) [reply adapter/config/parser.go]: no, removed and resolved

File: adapter/config/default_config.go (hunk 3 above)
Comment (CrowleyRajapakse) [reply adapter/config/default_config.go]: removed and resolved

File: test/integration/integration/utils/http/http.go (hunk 5 above)
Comment (CrowleyRajapakse) [reply test/integration/integration/utils/http/http.go:386]: yes, this only happen once in the very first test case, after that this flow doesn't hit

File: test/integration/integration/tests/deploy-api-to-different-listener.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// APIDifferentListener test
+var APIDifferentListener = suite.IntegrationTest{
+	ShortName:   "APIDifferentListener",
+	Description: "An API is deployed to a different listener other than default gateway listener",
+	Manifests:   []string{"tests/api-different-listener.yaml"},
+	Test: func(t *testing.T, suite *suite.IntegrationTestSuite) {
+		ns := "gateway-integration-test-infra"
+		gwAddr := kubernetes.WaitForAPIListenerAddress(t, suite.Client, suite.TimeoutConfig)
```
Comment (Amila-Rukshan) [inline test/integration/integration/tests/deploy-api-to-different-listener.go:39]: To test the api was added only to the defined listener, we need to invoke both listener addresses, isn't it?
You can call `WaitForGatewayAddress` and get the address with port `9095`, and when invoking it should give a `404`

File: adapter/internal/discovery/xds/server.go [hunk 12]
```diff
@@ -326,7 +331,9 @@ func updateXdsCacheOnAPIChange(oldLabels []string, newLabels []string) bool {
 	revisionStatus := false
 	// TODO: (VirajSalaka) check possible optimizations, Since the number of labels are low by design it should not be an issue
 	for _, newLabel := range newLabels {
-		listeners, clusters, routes, endpoints, apis := GenerateEnvoyResoucesForLabel(newLabel)
+		gateway := new(gwapiv1b1.Gateway)
+		gateway.Name = newLabel
+		listeners, clusters, routes, endpoints, apis := GenerateEnvoyResoucesForGateway(gateway, true, "APIController")
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply adapter/internal/discovery/xds/server.go]: resolved

File: helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml [hunk 13]
```diff
@@ -145,10 +145,12 @@ spec:
           image: {{ .Values.wso2.apk.dp.gatewayRuntime.deployment.router.image }}
           imagePullPolicy: {{ .Values.wso2.apk.dp.gatewayRuntime.deployment.router.imagePullPolicy }}
           ports:
-            - containerPort: 9095
+            - containerPort: {{ .Values.wso2.apk.dp.gateway.listener.port }}
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply helm-charts/templates/data-plane/gateway-components/gateway-runtime/gateway-runtime-deployment.yaml:148]: resolved

File: adapter/internal/discovery/xds/server.go [hunk 14]
```diff
@@ -411,22 +420,60 @@ func GenerateEnvoyResoucesForLabel(label string) ([]types.Resource, []types.Reso
...
+	var routesConfig []*routev3.RouteConfiguration
+
+	logger.LoggerXds.Debugf("Flow : %v", flow)
+
+	if flow == "GatewayController" {
+		listenerArray = oasParser.GetProductionListener(gateway)
+		envoyListenerConfigMap[gateway.Name] = listenerArray
+		routesConfig = envoyRouteConfigMap[gateway.Name]
+
+	} else if flow == "APIController" {
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply adapter/internal/discovery/xds/server.go]: resolved

File: adapter/internal/discovery/xds/server.go [hunk 15]
```diff
@@ -411,22 +420,60 @@ func GenerateEnvoyResoucesForLabel(label string) ([]types.Resource, []types.Reso
...
-	listenerArray, listenerFound := envoyListenerConfigMap[label]
-	routesConfig, routesConfigFound := envoyRouteConfigMap[label]
-	if !listenerFound && !routesConfigFound {
-		listenerArray, routesConfig = oasParser.GetProductionListenerAndRouteConfig(vhostToRouteArrayMap)
-		envoyListenerConfigMap[label] = listenerArray
-		envoyRouteConfigMap[label] = routesConfig
-	} else {
-		// If the routesConfig exists, the listener exists too
-		oasParser.UpdateRoutesConfig(routesConfig, vhostToRouteArrayMap)
+	var listenerArray []*listenerv3.Listener
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply adapter/internal/discovery/xds/server.go:428]: resolved, added it to the gatewayListener only

File: test/integration/integration/utils/kubernetes/helpers.go (hunk 4 above)
Comment (CrowleyRajapakse) [reply test/integration/integration/utils/kubernetes/helpers.go:75]: done, resolved

**Pull request closed without merging**